File: apps/backend/app/schemas/streams.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import Field

from app.schemas.registry import RegistrySchema


class StreamStartRequest(RegistrySchema):
    connection_id: UUID | None = None
    preferred_adapter: Literal["rtsp", "hls"] | None = None
    target_fps: float | None = Field(default=None, ge=0.1, le=60)
    decode_fps: float | None = Field(default=None, ge=0.5, le=60)
    transport: Literal["tcp", "udp"] | None = None
    max_frame_age_ms: int | None = Field(default=None, ge=50, le=10_000)


class StreamCameraRead(RegistrySchema):
    id: UUID
    camera_code: str
    camera_name: str
    department_id: UUID
    department_name: str
    district: str
    city: str | None
    latitude: float
    longitude: float
    vendor: str | None
    model: str | None
    camera_type: str
    ai_capabilities: list[str]


class StreamProfileRead(RegistrySchema):
    id: UUID
    name: str
    adapter_kind: str
    stream_role: str
    endpoint_display: str


class StreamMetricsRead(RegistrySchema):
    frames_received: int
    frames_dropped: int
    frames_sampled_out: int
    frames_dispatched: int
    stale_frames_dropped: int
    dropped_due_to_backpressure: int
    reconnect_count: int
    decoder_errors: int
    queue_depth: int
    source_fps: float | None
    decoded_fps: float
    processing_fps: float
    current_frame_age_ms: float | None
    average_frame_age_ms: float | None
    p95_frame_age_ms: float | None
    max_frame_age_ms: float | None
    last_frame_at: datetime | None
    last_dispatch_at: datetime | None
    clock_offset_ms: float | None
    resolution: str | None
    latency_estimate_ms: float | None
    bitrate_kbps: float | None
    latest_source_pts_seconds: float | None
    pts_timing_active: bool
    source_failover_count: int


class StreamSessionRead(RegistrySchema):
    id: UUID
    camera: StreamCameraRead
    profile: StreamProfileRead
    state: Literal[
        "created",
        "connecting",
        "connected",
        "streaming",
        "degraded",
        "reconnecting",
        "failed",
        "stopped",
    ]
    decoder_backend: str
    transport: str
    width: int
    height: int
    target_fps: float
    decode_fps: float
    buffer_capacity: int
    max_frame_age_ms: int
    metrics: StreamMetricsRead
    created_at: datetime
    state_changed_at: datetime
    connected_at: datetime | None
    stopped_at: datetime | None
    last_error_code: str | None
    last_error_message: str | None
    preview_url: str | None


class StreamSessionList(RegistrySchema):
    items: list[StreamSessionRead]
    total: int


class StreamCapabilitiesRead(RegistrySchema):
    available: bool
    decoder_backend: str | None
    decoder_source: str | None
    configured_backend: str
    hardware_decode_active: bool
    hardware_decode_reason: str
    gpu_zero_copy_active: bool
    latest_frame_semantics: bool
    batch_dispatch: bool
    max_active_sessions: int
    supported_source_types: list[str]


class AnalyticsDetectionRead(RegistrySchema):
    kind: Literal["object", "plate"]
    class_id: int
    class_name: str
    confidence: float
    x1: float
    y1: float
    x2: float
    y2: float
    plate_text: str | None
    ocr_confidence: float | None
    track_id: int | None = None


class CameraAnalyticsRead(RegistrySchema):
    camera_id: UUID
    stream_id: UUID
    frame_number: int
    observed_at: datetime
    status: str
    model: str
    device: str
    inference_ms: float
    detections: list[AnalyticsDetectionRead]
    routed_modules: list[str]
    error_message: str | None


class CameraAnalyticsList(RegistrySchema):
    items: list[CameraAnalyticsRead]
    total: int


class AnalyticsCapabilitiesRead(RegistrySchema):
    enabled: bool
    status: str
    consumer_attached: bool
    model: str | None
    device: str | None
    reason: str | None
    routes: list[str]


class StreamStateCountsRead(RegistrySchema):
    created: int
    connecting: int
    connected: int
    streaming: int
    degraded: int
    reconnecting: int
    failed: int
    stopped: int


class StreamAggregateMetricsRead(RegistrySchema):
    active_streams: int
    offline_streams: int
    degraded_streams: int
    reconnecting_streams: int
    average_decoded_fps: float
    average_processing_fps: float
    average_latency_ms: float
    total_frames_received: int
    total_frames_dropped: int
    total_reconnects: int
    scheduler_queue_depth: int
    ai_consumer_attached: bool
    worker_cpu_percent: float
    worker_memory_mb: float
    worker_processes: int
    gpu_decode_utilization_percent: float | None
    network_receive_mbps: float | None
    states: StreamStateCountsRead


def stream_session_read(snapshot: object) -> StreamSessionRead:
    return StreamSessionRead.model_validate(snapshot)

File: apps/frontend/src/types/streams.ts
export type ProcessingStreamState =
  | "created"
  | "connecting"
  | "connected"
  | "streaming"
  | "degraded"
  | "reconnecting"
  | "failed"
  | "stopped";

export interface ProcessingStreamMetrics {
  frames_received: number;
  frames_dropped: number;
  frames_sampled_out: number;
  frames_dispatched: number;
  stale_frames_dropped: number;
  dropped_due_to_backpressure: number;
  reconnect_count: number;
  decoder_errors: number;
  queue_depth: number;
  source_fps: number | null;
  decoded_fps: number;
  processing_fps: number;
  current_frame_age_ms: number | null;
  average_frame_age_ms: number | null;
  p95_frame_age_ms: number | null;
  max_frame_age_ms: number | null;
  last_frame_at: string | null;
  last_dispatch_at: string | null;
  clock_offset_ms: number | null;
  resolution: string | null;
  latency_estimate_ms: number | null;
  bitrate_kbps: number | null;
  latest_source_pts_seconds?: number | null;
  pts_timing_active?: boolean;
  source_failover_count?: number;
}

export interface ProcessingStreamCamera {
  id: string;
  camera_code: string;
  camera_name: string;
  department_id: string;
  department_name: string;
  district: string;
  city: string | null;
  latitude: number;
  longitude: number;
  vendor: string | null;
  model: string | null;
  camera_type: string;
  ai_capabilities?: string[];
}

export interface ProcessingStreamSession {
  id: string;
  camera: ProcessingStreamCamera;
  profile: {
    id: string;
    name: string;
    adapter_kind: string;
    stream_role: string;
    endpoint_display: string;
  };
  state: ProcessingStreamState;
  decoder_backend: string;
  transport: string;
  width: number;
  height: number;
  target_fps: number;
  decode_fps: number;
  buffer_capacity: number;
  max_frame_age_ms: number;
  metrics: ProcessingStreamMetrics;
  created_at: string;
  state_changed_at: string;
  connected_at: string | null;
  stopped_at: string | null;
  last_error_code: string | null;
  last_error_message: string | null;
  preview_url: string | null;
}

export interface ProcessingStreamSessionList {
  items: ProcessingStreamSession[];
  total: number;
}

export interface ProcessingStreamStart {
  connection_id?: string;
  preferred_adapter?: "rtsp" | "hls";
  target_fps?: number;
  decode_fps?: number;
  transport?: "tcp" | "udp";
  max_frame_age_ms?: number;
}

export interface StreamAggregateMetrics {
  active_streams: number;
  offline_streams: number;
  degraded_streams: number;
  reconnecting_streams: number;
  average_decoded_fps: number;
  average_processing_fps: number;
  average_latency_ms: number;
  total_frames_received: number;
  total_frames_dropped: number;
  total_reconnects: number;
  scheduler_queue_depth: number;
  ai_consumer_attached: boolean;
  worker_cpu_percent: number;
  worker_memory_mb: number;
  worker_processes: number;
  gpu_decode_utilization_percent: number | null;
  network_receive_mbps: number | null;
  states: Record<ProcessingStreamState, number>;
}

export interface StreamCapabilities {
  available: boolean;
  decoder_backend: string | null;
  decoder_source: string | null;
  configured_backend: string;
  hardware_decode_active: boolean;
  hardware_decode_reason?: string;
  gpu_zero_copy_active: boolean;
  latest_frame_semantics: boolean;
  batch_dispatch: boolean;
  max_active_sessions: number;
  supported_source_types: string[];
}

export interface AnalyticsDetection {
  kind: "object" | "plate";
  class_id: number;
  class_name: string;
  confidence: number;
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  plate_text: string | null;
  ocr_confidence: number | null;
  track_id: number | null;
}

export interface CameraAnalytics {
  camera_id: string;
  stream_id: string;
  frame_number: number;
  observed_at: string;
  status: string;
  model: string;
  device: string;
  inference_ms: number;
  detections: AnalyticsDetection[];
  routed_modules: string[];
  error_message: string | null;
}

export interface CameraAnalyticsList {
  items: CameraAnalytics[];
  total: number;
}

export interface AnalyticsCapabilities {
  enabled: boolean;
  status: "disabled" | "initializing" | "active" | "unavailable" | string;
  consumer_attached: boolean;
  model: string | null;
  device: string | null;
  reason: string | null;
  routes: string[];
}
